consensus/qbft/core/roundchange.go

```diff
@@ -0,0 +1,291 @@
+
+	// Sign message
+	encodedPayload, err := roundChange.EncodePayloadForSigning()
+	if err != nil {
+		logger.Error("QBFT: Failed to encode round-change message", "msg", roundChange, "err", err)
+		return
+	}
+	signature, err := c.backend.Sign(encodedPayload)
+	if err != nil {
+		logger.Error("QBFT: Failed to sign round-change message", "msg", roundChange, "err", err)
+		return
+	}
+	roundChange.SetSignature(signature)
+
+	// Add justification
+	if c.QBFTPreparedPrepares != nil {
+		roundChange.Justification = c.QBFTPreparedPrepares
+		logger.Info("QBFT: On RoundChange", "justification", roundChange.Justification)
+	}
+
+	// RLP-encode message
+	data, err := rlp.EncodeToBytes(roundChange)
+	if err != nil {
+		logger.Error("QBFT: Failed to encode round-change message", "msg", roundChange, "err", err)
+		return
+	}
+
+	logger.Info("QBFT: broadcast round-change message", "msg", roundChange)
+	// Broadcast RLP-encoded message
+	if err = c.backend.Broadcast(c.valSet, roundChange.Code(), data); err != nil {
+		logger.Error("QBFT: Failed to broadcast message", "msg", roundChange, "err", err)
+		return
+	}
+}
+
+func (c *core) handleRoundChange(roundChange *message.RoundChange) error {
+	logger := c.logger.New("state", c.state)
+
+	logger.Info("QBFT: handleRoundChange", "m", roundChange)
+
+	view := roundChange.View()
+	currentRound := c.currentView().Round
+
+	// Add the ROUND CHANGE message to its message set and return how many
+	// messages we've got with the same round number and sequence number.
+	if view.Round.Cmp(currentRound) >= 0 {
+		var prepareMessages []*message.SignedPreparePayload = nil
+		var pr *big.Int = nil
+		var pb *types.Block = nil
+		if roundChange.PreparedRound != nil && roundChange.PreparedBlock != nil && roundChange.Justification != nil && len(roundChange.Justification) > 0 {
+			prepareMessages = roundChange.Justification
+			pr = roundChange.PreparedRound
+			pb = roundChange.PreparedBlock
+		}
+		err := c.roundChangeSet.Add(view.Round, roundChange, pr, pb, prepareMessages, c.QuorumSize())
+		if err != nil {
+			logger.Warn("Failed to add round change message", "msg", roundChange, "err", err)
+			return err
+		}
+	}
+
+	num := c.roundChangeSet.higherRoundMessages(currentRound)
+	currentRoundMessages := c.roundChangeSet.getRCMessagesForGivenRound(currentRound)
+	log.Info("QBFT: handleRoundChange count", "higherRoundMsgs", num, "currentRoundMsgs", currentRoundMessages)
+	if num == c.valSet.F()+1 {
+		newRound := c.roundChangeSet.getMinRoundChange(currentRound)
+		logger.Trace("Starting new Round", "round", newRound)
+		c.startNewRound(newRound)
+		c.broadcastRoundChange(newRound)
+	} else if currentRoundMessages >= c.QuorumSize() && c.IsProposer() && c.current.preprepareSent.Cmp(currentRound) < 0 {
+		_, proposal := c.highestPrepared(currentRound)
+		if proposal == nil {
+			proposal = c.current.pendingRequest.Proposal
```

Comment:
I have seen some flaky panic here because `c.current.pendingRequest` is `nil`.

Should we extend if check? 

```go
if proposal == nil && c.current.pendingRequest != nil
```